File: src/leetcode/109.有序链表转换二叉搜索树.php
<?php
require_once 'ds.php';

use PHPUnit\Framework\TestCase;

/*
 * @lc app=leetcode.cn id=109 lang=php
 *
 * [109] 有序链表转换二叉搜索树
 */

// @lc code=start
/**
 * Definition for a singly-linked list.
 * class ListNode {
 *     public $val = 0;
 *     public $next = null;
 *     function __construct($val = 0, $next = null) {
 *         $this->val = $val;
 *         $this->next = $next;
 *     }
 * }
 */

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     public $val = null;
 *     public $left = null;
 *     public $right = null;
 *     function __construct($val = 0, $left = null, $right = null) {
 *         $this->val = $val;
 *         $this->left = $left;
 *         $this->right = $right;
 *     }
 * }
 */
class Solution
{

    /**
     * @param ListNode $head
     * @return TreeNode
     */
    function sortedListToBST($head)
    {
        $arr = [];
        while ($head) {
            $arr[] = $head->val;
            $head = $head->next;
        }
        return $this->BST($arr);
    }

    function BST(array $arr)
    {
        if (!$arr) return null;
        $mid = (int)(count($arr) / 2);
        $root = new TreeNode($arr[$mid]);
        $root->left = $this->BST(array_slice($arr, 0, $mid));
        $root->right = $this->BST(array_slice($arr, $mid + 1));
        return $root;
    }
}
// @lc code=end

File: src/leetcode/109.有序链表转换二叉搜索树.go
package leetcode

/*
 * @lc app=leetcode.cn id=109 lang=golang
 *
 * [109] 有序链表转换二叉搜索树
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func sortedListToBST(head *ListNode) *TreeNode {
	var a []int
	for head != nil {
		a = append(a, head.Val)
		head = head.Next
	}
	return BST109(a)

}
func BST109(a []int) *TreeNode {
	if len(a) == 0 {
		return nil
	}
	mid := len(a) / 2
	root := &TreeNode{Val: a[mid]}
	root.Left = BST109(a[:mid])
	root.Right = BST109(a[mid+1:])
	return root
}

// @lc code=end
